### ISSUE TITLE
ACE filter function creates noisy images due to data processing

### ISSUE THREAD

**Issue by FahadahmedK** [guillaumejs2403/ACE#3]
Hi,

I trained the DDPM generator using your code on the CelebA dataset. I noticed that the filter function in main.py creates a noisy counterfactual explanation due to the following data processing step:

ce = (pe.detach() - 0.5) / 0.5

I found this problem when I performed some sanity checks, for instance, by running the ace filter function without any attack iterations. When I commented out the above data processing step, however, I was able to create a way cleaner image. Would you have any idea why this could be happening?

I am also attaching the noisy (created with no changes to the code) and the clean (created with data processing step commented out) counterfactuals. Kindly note that these counterfactuals are produced without any attack. Thank you.
![noisy_ce](https://github.com/guillaumejs2403/ACE/assets/36238447/eefc4eec-b419-4764-ab4e-fd7ebf7c7bd8)
![ce](https://github.com/guillaumejs2403/ACE/assets/36238447/ee6df6f1-07d9-45a9-8b70-657170c45cc7)

**Comment by guillaumejs2403** [comment]
Hi, 

With this information, it is a little hard to understand what is happening. But I'll give it a try. 
1) When you say you are using no attack, do you mean you are using the flag `--attack_method=None`? If you are filtering the image, are you sure your model is working fine? Could you show me the original image?
2) How are you debugging the code? Looking at it, if you commend [L157](https://github.com/guillaumejs2403/ACE/main.py#L157), it should not work since in L157 it is the first time the variable `ce` is instantiated. Thus, [L168](https://github.com/guillaumejs2403/ACE/main.py#L168) should get an error. My guess is that you are generating the `ce` variable a priori and the function is working fine.
4) Are you looking at the correct outputs of the function? I guess you do :P, but just confirm it - first is the `ce`, second `pe`, third `noise`, and last `mask`.

Kindly,
Guillaume

**Comment by FahadahmedK** [comment]
Hi,

Thank you for your answer, first of all. I shall try to answer your questions and explain my issues further with as many details as I can:

1. To not attack at all, I am merely setting the _attack_iterations_ parameter to 0. I wanted to check why I am only able to retrieve noisy counterfactuals (I will further explain in 2 how I exactly debugged the code). I used _diffusion.q\_sample()_ and _diffusion.p\_sample()_ methods to check if the generator I trained using your code worked well in the first place - and it did. Below I have attached the original image, along with the corrupted version using q.sample(), and the filtered version using p.sample() methods.
![img_forward_reverse](https://github.com/guillaumejs2403/ACE/assets/36238447/edebb012-42c5-4e6f-b5c4-3ef9d6e9634f)

2. I am basically saving all the outputs for debugging including boolmask, pe_mask, ce_mask, pe, and ce. I did not just comment out [L157](https://github.com/guillaumejs2403/ACE/blob/0052a88c68623a8711315d393289d7d245cb99db/main.py#L157) but I changed it slightly. I removed the rescaling step and changed it to _ce = pe.detach()_ and also commented out [L187](https://github.com/guillaumejs2403/ACE/blob/0052a88c68623a8711315d393289d7d245cb99db/main.py#L187), of course. I ran the code with the aforementioned settings and found out that it produced a cleaner image. My finding was that when this rescaling step is performed, the filtering process produces a noisy image as I showed to you in my first post.

3. As far as I know, I think I am looking at the correct outputs because I save all the outputs directly within the function for visual inspection.

Kind regards,

Fahad

**Comment by guillaumejs2403** [comment]
From what you say, I suspect that you trained your model without any normalization. Please try using the model I provided as another sanity check ([Link](https://drive.google.com/file/d/1OqjWns4NSu6AiKkOnpUOjUHzA8sQlaOA/view?usp=sharing))

**Comment by guillaumejs2403** [comment]
Closed due to inactivity
